webhooks/virtualmachine/mutation/scsi_controller.go

```diff
@@ -0,0 +1,252 @@
+// © Broadcom. All Rights Reserved.
+// The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package mutation
+
+import (
+	apierrorsutil "k8s.io/apimachinery/pkg/util/errors"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha5"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	vmopv1util "github.com/vmware-tanzu/vm-operator/pkg/util/vmopv1"
+)
+
+// AddControllersForVolumes mutates the VM to add any controllers to
+// the spec if any of the specified volumes cannot be accommodate on
+// the existing controllers. If one of the current controllers
+// (spec.hardware.controllers) is suitable for the volume (e.g.,
+// matches the sharing mode), we check if there are any open slots. If
+// none are found, we add a controller.
+func AddControllersForVolumes(
+	ctx *pkgctx.WebhookRequestContext,
+	client ctrlclient.Client,
+	vm *vmopv1.VirtualMachine) (bool, error) {
+
+	var (
+		errs        []error
+		err         error
+		scsiMutated bool
+	)
+	scsiMutated, err = AddSCSIControllersForVolumes(ctx, client, vm)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// TODO: handle other controller types.
```

Comment:
I have given this instruction multiple times to others and in internal chat, so please make sure to adhere to it. All `TODO` entries should follow the pattern `TODO(FeatureName)` in order to make it easy to find outstanding items related to a given feature. If there is no feature, use your GitHub ID. In this case there _is_ a feature. Please follow directions.